[14.0][ADD] sale_unreconciled

AaronHForgeFlow opened the pull request:
This module adds a new fields "Unreconciled" on Sales Orders, that allows
to find SO's with unreconciled journal items related.

This module allows to reconcile those SO in a single click. In accounting
settings users will be able to set up a specific account for write-off.

cc @ForgeFlow

AaronHForgeFlow committed:
[14.0][ADD] sale_unreconciled

ChrisOForgeFlow commented on sale_unreconciled/models/account_move_line.py:
```diff
@@ -0,0 +1,69 @@
+        counter_part["debit"] = write_off_vals["credit"]
+        counter_part["credit"] = write_off_vals["debit"]
+        counter_part["amount_currency"] = -write_off_vals["amount_currency"]
+        counter_part["account_id"] = (counterpart_account.id,)
```
could this throw up a singleton problem?

AaronHForgeFlow replied:
Perhaps it could, I am adding a warning message just in case

ChrisOForgeFlow commented on sale_unreconciled/models/sale_order.py:
```diff
@@ -0,0 +1,142 @@
+             everything is reconciled or that the related accounts do not
+             allow reconciliation""",
+    )
+    is_delivered = fields.Boolean(search="_search_is_delivered")
```
I don't know if this feature is adequate in this module, because I would understand that it has more to do with the stock than with the reconciliation of the sales order.

AaronHForgeFlow replied:
The thing is, this field was part of a previous Odoo version and it was used by this module. But I agree it should be out of this module.

AaronHForgeFlow replied:
The field is removed now.

ChrisOForgeFlow commented on sale_unreconciled/models/sale_order.py:
```diff
@@ -0,0 +1,142 @@
+            }
+
+    def action_done(self):
+        if self.unreconciled:
```
this should be refactored to iterate over self, as in the original implementation there is no self.ensure_one ()

AaronHForgeFlow replied:
good point! done :)

ChrisOForgeFlow commented on sale_unreconciled/models/company.py:
```diff
@@ -0,0 +1,19 @@
+        help="Write-off account to reconcile Unreconciled Sale Orders",
+    )
+
+    sale_reconcile_journal_id = fields.Many2one("account.journal")
```
A string should be added in the definition, since the field is visible for the configuration

AaronHForgeFlow replied:
If I don't add a string, it shows as "Sale Reconcile Journal". It think it is correct but I don't mind to put another string if you suggest a different thing.

ChrisOForgeFlow commented on sale_unreconciled/views/sale_order_view.xml:
```diff
@@ -0,0 +1,59 @@
+                <filter
+                    name="finance_review"
+                    string="To Review By Finance"
+                    domain="[('unreconciled','=', True), ('invoice_status', '=', 'invoiced'), ('is_delivered', '=', True)]"
```
This sounds like a more specific customization

AaronHForgeFlow replied:
Actually, I think this filter would be the most useful utility in this module, as long as the "Unreconciled " filter will show sales that are unreconciled and they are correct. In any case I am removing this because the field is_delivered is removed as well.

AaronHForgeFlow replied:
Removed the filter, it is a custom thing that is not necessary to be in the module

AaronHForgeFlow committed:
[FIX] sale_unreconciled: do not create empty write-offs
[FIX] sale_unreconciled: ensure the accounting settigs are correct
[FIX] sale_unreconciled: assign sale order to the write-off created

AaronHForgeFlow committed:
[14.0][FIX] sale_unreconciled: reconcile exit condition

AaronHForgeFlow committed:
[IMP] sale_unreconciled: allow companies to decide automatic reconcile or not

AaronHForgeFlow committed:
[FIX]sale_unreconciled: reconcile by currency

AaronHForgeFlow committed:
[IMP] sale_unreconciled: add tolerance parameter

AaronHForgeFlow committed:
[FIX] sale_unreconciled: if journals have no residual but are full reconcile then assign a matching number
[IMP] sale_unreconciled: use the amount resiudal as condition for being unreconciled

AaronHForgeFlow committed:
[FIX] sale_unreconciled: multicompany for get unreconciled base domain

AaronHForgeFlow committed:
[FIX] sale_unreconciled: action to see unreconciled journal items has to consider the company of the SO

AaronHForgeFlow committed:
[FIX] sale_unreconciled: account domain

AaronHForgeFlow committed:
[IMP] sale_unreconciled: reconcile by groups

AaronHForgeFlow committed:
[IMP] sale_unreconciled: reconcile by groups

bosd commented:
Could you recreate a runboat?

AaronHForgeFlow commented:
Just rebased to see if runboat is recreated

bosd commented on sale_unreconciled/views/res_config_settings_view.xml:
```diff
@@ -0,0 +1,62 @@
+                                    groups="base.group_multi_company"
+                                />
+                            <div class="text-muted">
+                                Reconcile S O upon locking
```
```suggestion
                                Reconcile SO upon locking
```

bosd commented on sale_unreconciled/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,5 @@
+This module adds a new fields "Unreconciled" on Sales Orders, that allows
```
```suggestion
This module adds a new field "Unreconciled" on Sales Orders, that allows
```

bosd commented on sale_unreconciled/readme/USAGE.rst:
```diff
@@ -0,0 +1,6 @@
+Accountants will be able to find a filters in Sale Orders that shows
```
```suggestion
Accountants will be able to find a filter in Sale Orders that shows
```

bosd commented on sale_unreconciled/readme/USAGE.rst:
```diff
@@ -0,0 +1,6 @@
+to those outstanding journal items.
+
+Locking the SO will automatically reconcile the outstanding balance for the
+stock iterim accounts.
```
```suggestion
stock interim accounts.
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+        cls.partner = cls.partner_obj.create({"name": "Test Vendor"})
+        # Create standard product:
+        cls.product = cls.product_obj.create(
+            {"name": "saled Product", "type": "product"}
```
```suggestion
            {"name": "sold Product", "type": "product"}
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+        )
+        cls.product_to_reconcile = cls.product_obj.create(
+            {
+                "name": "saled Product (To reconcile)",
```
```suggestion
                "name": "sold Product (To reconcile)",
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+
+    def test_04_sale_mrp_anglo_saxon(self):
+        """Test sale order for kit, deliver and invoice and ensure
+        the iterim account is balanced and COGS is hit
```
```suggestion
        the interim account is balanced and COGS is hit
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+            sum(aml_output.mapped("credit")), 150, "GDNI missing or mismatching"
+        )
+        # Now checking that the stock entries for the finished product are
+        # created and not the component (decide if that is derisable)
```
```suggestion
        # created and not the component (decide if that is desirable)
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+        so.with_context(force_confirm_sale_order=True).action_confirm()
+        self._do_picking(so.picking_ids, fields.Datetime.now())
+        self.assertTrue(so.unreconciled)
+        # as long stock_dropshipping is not dependency, I force the SO to be in
```
```suggestion
        # as long stock_dropshipping is not an dependency, I force the SO to be in
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+            [("move_id.ref", "=", incoming_name)]
+        )
+        incoming_ji.write({"sale_line_id": so.order_line[0], "sale_order_id": so.id})
+        # then I lock the so to force reconciliation
```
```suggestion
        # then lock the SO to force reconciliation
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+        so._compute_unreconciled()
+        self.assertFalse(so.unreconciled)
+        # the SO is reconciled and the stock interim deliverd account is not
+        # reconciled yet
```
```suggestion
        #  yet reconciled
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+        so.action_done()
+        so._compute_unreconciled()
+        self.assertFalse(so.unreconciled)
+        # we check all the journals for the so have the same company
```
```suggestion
        # check all the journals for the so have the same company
```

bosd commented on sale_unreconciled/tests/test_sale_unreconciled.py:
```diff
@@ -0,0 +1,523 @@
+        self._do_picking(so.picking_ids, fields.Datetime.now())
+        # Do not create invoices to force discrepancy
+        so.action_done()
+        # we check all the journals are balanced by product
```
```suggestion
        # check if all the journals are balanced by product
```

bosd reviewed (commented):
LGTM, Minor nitpicking on grammar.

AaronHForgeFlow committed:
[IMP] sale_unreconciled: text

AaronHForgeFlow committed:
[IMP]sale_unreconciled: reconcile cancelled SO

bosd commented:
@alexis-via you might find this pr interesting.
